core/src/main/java/hudson/util/Retrier.java

```diff
@@ -0,0 +1,177 @@
+package hudson.util;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.concurrent.Callable;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+/**
+ * This class implements a process of doing some action repeatedly synchronously until it is performed successfully.
+ * You can set the number of attempts, the action to perform, the milliseconds to wait for, the definition of success,
+ * the exceptions that are considered as a failed action, but not an unexpected exception in the action and also the
+ * listener to manage the expected exceptions happened, just in case it is helpful.
+ * @param <V> The return type of the action to perform.
+ */
+public class Retrier <V>{
```

Comment:
I think we should mark this `@Restricted(NoExternalUse.class)` at least for some time to let this soak, so that we can possibly fix/improve the design in future releases before it is possibly relaxed and made available to all plugins out there?